module #2. REST API Basics/rest-api-basic/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,241 @@
+            }
+            log.debug("[GiftCertificateService.findByName()] GiftCertificate received from database: [{}], for name:[{}]"
+                    , certificates, name);
+            return certificates;
+        }
+    }
+
+    @Override
+    public List<GiftCertificateDTO> findAll() {
+        List<GiftCertificateDTO> certificates = giftCertificateRepository.findAll()
+                .stream()
+                .flatMap(Collection::stream)
+                .map(certificateMappingService::mapToDto)
+                .collect(Collectors.toList());
+        return getCertificateDTOSWithTags(certificates);
+    }
+
+    @Override
+    public List<GiftCertificateDTO> findAllWithParams(QueryParams queryParams) {
+        if (queryParams == null) {
+            throw new IllegalArgumentException();
+        }
+        List<GiftCertificateDTO> certificates = giftCertificateRepository.findAllWithParams(queryParams)
+                .stream()
+                .flatMap(Collection::stream)
+                .map(certificateMappingService::mapToDto)
+                .collect(Collectors.toList());
+        return getCertificateDTOSWithTags(certificates);
+    }
+
+    @Transactional
+    @Override
+    public void update(Long id, GiftCertificateDTO giftCertificateDTO) {
+        Validate.notNull(giftCertificateDTO, "GiftCertificateDTO can't be Null");
+        if (id == null || id < 1) {
+            log.error("[GiftCertificateService.update()] An exception occurs: given ID:[{}]" +
+                    " can't be less than zero or null", id);
+            throw new IllegalArgumentException("Given ID can't be less than zero or null");
+        }
+
+        GiftCertificate giftCertificate = certificateMappingService.mapFromDto(giftCertificateDTO);
+        giftCertificate.setLastUpdateDate(LocalDateTime.now());
+        transactionTemplate.execute(new TransactionCallbackWithoutResult() {
+            @Override
+            protected void doInTransactionWithoutResult(TransactionStatus transactionStatus) {
+                try {
+                    giftCertificateRepository.update(id, giftCertificate);
+                    attachAndSaveTags(giftCertificate, id);
+                    log.debug("[GiftCertificateService.update()] GiftCertificate with ID:[{}] updated.", id);
+                } catch (Exception e) {
+                    transactionStatus.setRollbackOnly();
+                    log.error("[GiftCertificateService.update()] rollback !!!");
+                }
+            }
+        });
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        if (id == null || id < 1) {
+            log.error("[GiftCertificateService.deleteById()] " +
+                    "An exception occurs: id:[{}] can't be less than zero or null", id);
+            throw new IllegalArgumentException("GiftCertificate.id can't be less than zero or null");
+        }
+        if (giftCertificateRepository.findById(id).isEmpty()) {
```

Comment:
it is better to use exists method which will return only boolean value, because in this case there is no need for you to return all data about this entity just to know if it is not null.